pkg/k8s/client/object_tracker.go

```diff
@@ -0,0 +1,617 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package client
+
+// Some parts of this code is originally from k8s.io/client-go/testing
+// Copyright 2015 The Kubernetes Authors.
+
+import (
+	"encoding/json"
+	"fmt"
+	"log/slog"
+	"reflect"
+	"strconv"
+	"strings"
+
+	"github.com/cilium/statedb"
+	"github.com/cilium/statedb/index"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/selection"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/apimachinery/pkg/util/strategicpatch"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/testing"
+	"k8s.io/client-go/util/jsonpath"
+
+	"github.com/cilium/cilium/pkg/k8s/testutils"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+)
+
+const (
+	logfieldGVR             = "gvr" //  GroupVersionResource
+	logfieldClientset       = "clientset"
+	logfieldResourceVersion = "resourceVersion"
+	logfieldFieldSelector   = "fieldSelector"
+)
+
+// statedbObjectTracker implements [testing.ObjectTracker] using a StateDB table.
+// This allows for proper implementation of Watch() that respects the ResourceVersion
+// allowing multiple Watch() calls in parallel.
+type statedbObjectTracker struct {
+	domain  string
+	log     *slog.Logger
+	db      *statedb.DB
+	scheme  testing.ObjectScheme
+	decoder runtime.Decoder
+	tbl     statedb.RWTable[object]
+}
+
+func newStateDBObjectTracker(db *statedb.DB, log *slog.Logger) (*statedbObjectTracker, error) {
+	tbl, err := statedb.NewTable("k8s-object-tracker", objectIndex)
+	if err != nil {
+		return nil, err
+	}
+	if err := db.RegisterTable(tbl); err != nil {
+		return nil, err
+	}
+	return &statedbObjectTracker{
+		log:     log,
+		db:      db,
+		tbl:     tbl,
+		scheme:  testutils.Scheme,
+		decoder: testutils.Decoder(),
+	}, nil
+}
+
+type object struct {
+	objectId
+	o runtime.Object
+}
+
+type objectId struct {
+	types.NamespacedName
+	gvr schema.GroupVersionResource
+
+	// domain to which this objects belongs, e.g. this is used
+	// to differentiate between the slim and kubernetes clientsets
+	// and avoid mixing them up since they have the same [gvr]
+	domain string
```

Comment:
This is a bit clunky. The fake client uses "*" for everything except "kubernetes" for the core kubernetes clientsets. The reason for this is the overlap between slim and kubernetes and without this we'd mix them up. Could also have completely separate StateDB tables for each "clientset", but I kind of like just having one that is easy to inspect.